Development/nmos/constraints.cpp

```diff
@@ -0,0 +1,256 @@
+#include "nmos/constraints.h"
+
+#include <algorithm>
+#include <set>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+#include "cpprest/basic_utils.h" // for utility::us2s
+#include "cpprest/json_utils.h"
+#include "nmos/capabilities.h"
+#include "nmos/json_fields.h"
+#include "nmos/rational.h"
+
+namespace nmos
+{
+    namespace experimental
+    {
+        namespace detail
+        {
+            bool constraint_value_less(const web::json::value& lhs, const web::json::value& rhs)
+            {
+                if (nmos::is_rational(lhs) && nmos::is_rational(rhs))
+                {
+                    if (nmos::parse_rational(lhs) < nmos::parse_rational(rhs))
+                    {
+                        return true;
+                    }
+                    return false;
+                }
+                return lhs < rhs;
+            }
+        }
+
+        web::json::value get_intersection(const web::json::array& lhs_, const web::json::array& rhs_)
+        {
+            std::set<web::json::value, decltype(&detail::constraint_value_less)> lhs(lhs_.begin(), lhs_.end(), &detail::constraint_value_less);
+            std::set<web::json::value, decltype(&detail::constraint_value_less)> rhs(rhs_.begin(), rhs_.end(), &detail::constraint_value_less);
+
+            std::vector<web::json::value> v(std::min(lhs.size(), rhs.size()));
+
+            const auto it = std::set_intersection(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), v.begin(), detail::constraint_value_less);
+            v.resize(it - v.begin());
+
+            return web::json::value_from_elements(v);
```

Comment:
I haven't got a perfect answer yet, but this feels inefficient as we copy values three times. Will have a think.

Reply:
What do you think about applying your idea from the comment above? Would using a `back_inserter` into `web::json::value result` save us one copy?

Reply:
Hmm, have we been overthinking this? Can we just create an empty set with constraint_value_less, then insert all of the lhs, then insert all of the rhs, then do  value_from_elements?

Reply:
In this case we get values that appear at least in one of these arrays instead of values that appear in both arrays, isn't it?

Reply:
Yeah, not sure what I was thinking 😅